bundles/io/org.eclipse.smarthome.io.voice/src/main/java/org/eclipse/smarthome/io/voice/text/ExpressionIdentifier.java

```diff
@@ -32,40 +33,40 @@ public ExpressionIdentifier(AbstractRuleBasedInterpreter interpreter) {
      * Constructs a new instance.
      *
      * @param interpreter the interpreter it belongs to. Used for dynamically fetching item name tokens
-     * @param excludes tokens that should not occur for this expression to match
+     * @param stopper Expression that should not match, if the current token should be accepted as identifier
      */
-    public ExpressionIdentifier(AbstractRuleBasedInterpreter interpreter, HashSet<String> excludes) {
+    public ExpressionIdentifier(AbstractRuleBasedInterpreter interpreter, Expression stopper) {
         super();
         this.interpreter = interpreter;
-        this.excludes = excludes == null ? new HashSet<String>() : new HashSet<String>(excludes);
+        this.stopper = stopper;
     }
 
     @Override
-    ASTNode parse(TokenList list) {
+    ASTNode parse(ResourceBundle language, TokenList list) {
         ASTNode node = new ASTNode();
-        HashSet<String> tokens = interpreter.getIdentifierTokens();
-        String head = list.head();
-        node.setSuccess(head != null /* && tokens.contains(head) */ && !excludes.contains(head));
+        node.setSuccess(list.size() > 0 && (stopper == null || !stopper.parse(language, list).isSuccess()));
         if (node.isSuccess()) {
             node.setRemainingTokens(list.skipHead());
-            node.setValue(head);
+            node.setValue(list.head());
             node.setChildren(new ASTNode[0]);
             generateValue(node);
         }
         return node;
     }
 
     @Override
-    boolean collectFirsts(HashSet<String> firsts) {
-        HashSet<String> f = new HashSet<String>(interpreter.getIdentifierTokens());
-        f.removeAll(excludes);
+    boolean collectFirsts(ResourceBundle language, HashSet<String> firsts) {
+        HashSet<String> f = new HashSet<String>(interpreter.getAllItemTokens(language.getLocale()));
```

Comment:
You've had to make a lot of code changes to pass the ResourceBundle around just for this one line of code. Have you considered just making the language a property of the interpreter or something? Like interpreter.setCurrentLanguage() and interpreter.getCurrentLanguage(), maybe? That might simplify a lot of method signatures, if it is general enough for you.